leo/span/src/source_map.rs

```diff
@@ -0,0 +1,441 @@
+    /// number. If the source_file is empty or the position is located before the
+    /// first line, `None` is returned.
+    fn lookup_line(&self, pos: BytePos) -> Option<usize> {
+        match self.lines.binary_search(&pos) {
+            Ok(idx) => Some(idx),
+            Err(0) => None,
+            Err(idx) => Some(idx - 1),
+        }
+    }
+
+    /// Looks up the file's (1-based) line number and (0-based `CharPos`) column offset, for a
+    /// given `BytePos`.
+    fn lookup_file_pos(&self, pos: BytePos) -> (usize, CharPos) {
+        let chpos = self.bytepos_to_file_charpos(pos);
+        match self.lookup_line(pos) {
+            Some(a) => {
+                let line = a + 1; // Line numbers start at 1
+                let linebpos = self.lines[a];
+                let linechpos = self.bytepos_to_file_charpos(linebpos);
+                let col = chpos - linechpos;
+                assert!(chpos >= linechpos);
+                (line, col)
+            }
+            None => (0, chpos),
+        }
+    }
+
+    /// Returns contents of a `span` assumed to be within the given file.
+    fn contents_of_span(&self, span: Span) -> String {
+        let begin_pos = self.bytepos_to_file_charpos(span.lo).to_usize();
+        let end_pos = self.bytepos_to_file_charpos(span.hi).to_usize();
+        String::from_utf8_lossy(&self.src.as_bytes()[begin_pos..end_pos]).into_owned()
+    }
+}
+
+/// Detailed information on a `Span`.
+pub struct SpanLocation {
+    pub source_file: Rc<SourceFile>,
+    pub line_start: usize,
+    pub line_stop: usize,
+    pub col_start: usize,
+    pub col_stop: usize,
+}
+
+impl SpanLocation {
+    /// Returns a dummy location.
+    pub fn dummy() -> Self {
+        let dummy = "<dummy>".to_owned();
+        let span = Span::dummy();
+        Self {
+            source_file: Rc::new(SourceFile {
+                name: FileName::Custom(dummy.clone()),
+                src: dummy,
+                start_pos: span.lo,
+                end_pos: span.hi,
+                lines: Vec::new(),
+                multibyte_chars: Vec::new(),
+            }),
+            line_start: 0,
+            line_stop: 0,
+            col_start: 0,
+            col_stop: 0,
+        }
+    }
+}
+
+/// File / Line / Column information on a `BytePos`.
+pub struct LineCol {
+    /// Information on the original source.
+    pub source_file: Rc<SourceFile>,
+    /// The 1-based line number.
+    pub line: usize,
+    /// The (0-based) column offset into the line.
+    pub col: CharPos,
+}
+
+/// Normalizes the source code and records the normalizations.
+fn normalize_src(src: &mut String) {
+    remove_bom(src);
+    normalize_newlines(src);
+}
+
+/// Removes UTF-8 BOM, if any.
+fn remove_bom(src: &mut String) {
+    if src.starts_with('\u{feff}') {
+        src.drain(..3);
+    }
+}
+
+/// Replaces `\r\n` with `\n` in-place in `src`.
+///
+/// Returns error if there's a lone `\r` in the string.
```

Comment:
Since this is true, this affects #1721. I'm personally in favor of this change.

Reply:
@acoglio a'ing you so you are also aware of this change. Though tbf we didn't handle `\r` before, however, now this explicitly handles them.